docs/spec.md

```diff
@@ -2288,23 +2288,59 @@ Computes dot products between slices of `lhs` and slices of `rhs` and produces a
+  * `effective_lhs = lhs`.
+* `lhs_result_dimensions = [d for d in axes(effective_lhs) and d not in lhs_batching_dimensions and d not in lhs_contracting_dimensions]`.
 * `rhs_result_dimensions = [d for d in axes(rhs) and d not in rhs_batching_dimensions and d not in rhs_contracting_dimensions]`.
 * `result_batching_index + result_lhs_index + result_rhs_index = result_index`
   where `size(result_batching_index) = size(lhs_batching_dimensions)`,
   `size(result_lhs_index) = size(lhs_result_dimensions)` and
   `size(result_rhs_index) = size(rhs_result_dimensions)`.
-* `transposed_lhs = transpose(lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
-* `transposed_lhs_slice = slice(result_batching_index + result_lhs_index + [:, ..., :])`.
-* `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(lhs, lhs_contracting_dimensions))`.
+* `transposed_lhs = transpose(effective_lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
+* `transposed_lhs_slice = slice(transposed_lhs, result_batching_index + result_lhs_index + [:, ..., :])`.
+* `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(effective_lhs, lhs_contracting_dimensions))`.
 * `transposed_rhs = transpose(rhs, rhs_batching_dimensions + rhs_result_dimensions + rhs_contracting_dimensions)`.
-* `transposed_rhs_slice = slice(result_batching_index + result_rhs_index + [:, ..., :])`.
+* `transposed_rhs_slice = slice(transposed_rhs, result_batching_index + result_rhs_index + [:, ..., :])`.
 * `reshaped_rhs_slice = reshape(transposed_rhs_slice, dims(rhs, rhs_contracting_dimensions))`.
-* `dot_product = reduce(
-  inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
-  init_values=[0],
-  dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
-  body=lambda x, y: add(x, y))`.
+* For `is_non_quantized_tensor(lhs) and is_non_quantized_tensor(rhs)`:
+  * `dot_product = reduce(
+      inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)],
+      init_values=[0],
+      dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
+      body=lambda x, y: add(x, y))`.
+* For `is_quantized_tensor(lhs) and is_quantized_tensor(rhs)`:
+  * `integer_dot_product = reduce(
+      inputs=[multiply((reshaped_lhs_slice - zero_point(reshaped_lhs_slice)),
+                       (reshaped_rhs_slice - zero_point(reshaped_rhs_slice))],
+      init_values=[0],
+      dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
+      body=lambda x, y: add(x, y))`.
+  * `rounded_dot_product = round_nearest_even(integer_dot_product * (scale(reshaped_lhs_slice) * scale(reshape_rhs_slice) / scale(result)))`.
+  * `dot_product = clamp(storage_min(result), rounded_dot_product + zero_point(result), storage_max(result))`.
+* For `is_non_quantized_tensor(lhs) and is_quantized_tensor(rhs)`:
+  * `integer_dot_product = reduce(
+      inputs=[multiply((reshaped_lhs_slice - zero_point(reshaped_lhs_slice)) * scale(reshaped_lhs_slice),
```

Comment:
Not sure if I understand correctly here. Do we need to calculate scale and zero point for lhs? isn't it `is_non_quantized_tensor(lhs)`?

Reply:
Also, `(reshaped_lhs_slice - zero_point(reshaped_lhs_slice)) * scale(reshaped_lhs_slice` dequantize the integer back to float so it may be inappropriate to call it  `integer_dot_product `

Reply:
what about `is_quantized_tensor(lhs) and is_non_quantized_tensor(rhs)`?

Reply:
Hi @rewu93 

>> it may be inappropriate to call it integer_dot_product 

Will be fixed in next commit. 


>> what about is_quantized_tensor(lhs) and is_non_quantized_tensor(rhs)?

We have  constraint to disallow that. As per (C14),

```
 If the operation uses quantized tensors:
  * (C14) Either `is_quantized_tensor(lhs) and is_quantized_tensor(rhs)` or  `is_non_quantized_tensor(lhs) and is_quantized_tensor(rhs)`.
```

Reply:
Just try to be as explicit as possible to avoid confusion. What I meant is that since the hybrid kernel is evaluated as DRQ, instead of 
* `is_non_quantized_tensor(lhs) and is_quantized_tensor(rhs)`:
  * `integer_dot_product = reduce(
      inputs=[multiply((reshaped_lhs_slice - zero_point(reshaped_lhs_slice)) * scale(reshaped_lhs_slice),(reshaped_rhs_slice - zero_point(reshaped_rhs_slice)) * scale(reshaped_rhs_slice)],
      init_values=[0],
      dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
      body=lambda x, y: add(x, y))`.
  * `dot_product = integer_dot_product * scale(reshaped_lhs_slice) * scale(reshape_rhs_slice)`.

this should be 
* `is_non_quantized_tensor(lhs) and is_quantized_tensor(rhs)`:
  * `integer_dot_product = reduce(
      inputs=[multiply((reshaped_lhs_slice - zero_point(reshaped_lhs_slice)),(reshaped_rhs_slice - zero_point(reshaped_rhs_slice))],
      init_values=[0],
      dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
      body=lambda x, y: add(x, y))`.
  * `dot_product = integer_dot_product * scale(reshaped_lhs_slice) * scale(reshape_rhs_slice)`.

@loganchien I suppose we don't need to round and clamp for the hybrid case here

Reply:
Thanks @rewu93 for the fix. This is exactly the fix I have in mind. Please take a look.

Reply:
LGTM

Reply:
In my opinions, dynamic range quantization is somewhat unique, which is implemented only in TFLite CPU kernel in the world as far as I know (even all other TFLite variants (XNNPACK and so on) do not support dynamic range quantization kernels). IMO, in terms of adoptions, it is bit early to make dynamic range quantization kernels as a part of the language specification in StableHLO. As dynamic range quantization support in TFLite CPU kernel can be recognized as a legacy support, it seems like a fine choice to use custom ops to express dynamic range quantization for the time being. However, in the long run, there are some ideas too support dynamic range quantization in the graph level later. At the same time, this might be a marathon race to ask the ML hardware implementers to support dynamic range quantization kernels more effectively.

Reply:
Thanks @abattery for the elaborate feedback.

>> it seems like a fine choice to use custom ops for expressing dynamic range quantization kernels for the time being.

This seems like a reasonable choice. As you pointed out, `In the long run, there are some ideas to support dynamic range quantization in the graph level later.`, IMO, it might be reasonable to not include any hybrid op for now (which have the potential of getting deprecated in the long run) and instead use custom ops to express dynamic range quantization.

Anyways, I will be happy to understand any other opinion.

Reply:
If we can express dynamic range quantization using not a single StableHLO Op, but a composite set of StableHLO ops it would be fine. We can fallback to the existing TFLite implementation by pattern matching to it, however I think it's important it be expressible via StableHLO Ops, not just as a custom Op. There is a fair bit of production models depending on this behavior.

Reply:
Hi @nutsiepully 
Thanks for the information. We will explore more into the options.